dart/dynamics/ShapeFrame.cpp

```diff
@@ -0,0 +1,394 @@
+}
+
+//==============================================================================
+void ShapeFrame::setProperties(const ShapeFrame::UniqueProperties& properties)
+{
+  setShape(properties.mShape);
+}
+
+//==============================================================================
+const ShapeFrame::Properties ShapeFrame::getShapeFrameProperties() const
+{
+  return Properties(getEntityProperties(), mShapeFrameP, getAddonProperties());
+}
+
+//==============================================================================
+void ShapeFrame::copy(const ShapeFrame& other)
+{
+  if (this == &other)
+    return;
+
+  setProperties(other.getShapeFrameProperties());
+}
+
+//==============================================================================
+void ShapeFrame::copy(const ShapeFrame* other)
+{
+  if (nullptr == other)
+    return;
+
+  copy(*other);
+}
+
+//==============================================================================
+ShapeFrame& ShapeFrame::operator=(const ShapeFrame& other)
+{
+  copy(other);
+  return *this;
+}
+
+//==============================================================================
+void ShapeFrame::setShape(const ShapePtr& shape)
+{
+  if (nullptr == shape)
+  {
+    dtwarn << "[ShapeFrame::setShape] Attempting to add a nullptr as a "
+           << " shape of ShapeNode '" << getName()
+           << "', which is not allowed. "
+           << "The shape of ShapeNode '" << getName() << "' hasn't changed.\n";
+    return;
+  }
+
+  if (shape == mShapeFrameP.mShape)
+    return;
+
+  ShapePtr oldShape = mShapeFrameP.mShape;
+
+  mShapeFrameP.mShape = shape;
+
+  mShapeUpdatedSignal.raise(this, oldShape, mShapeFrameP.mShape);
+}
+
+//==============================================================================
+ShapePtr ShapeFrame::getShape()
+{
+  return mShapeFrameP.mShape;
+}
+
+//==============================================================================
+ConstShapePtr ShapeFrame::getShape() const
+{
+  return mShapeFrameP.mShape;
+}
+
+//==============================================================================
+VisualAddon* ShapeFrame::getVisualAddon(const bool createIfNull)
+{
+  VisualAddon* visualData = getVisualAddon();
+
+  if (createIfNull && nullptr == visualData)
+    return createVisualAddon();
+
+  return visualData;
+}
+
+//==============================================================================
+CollisionAddon* ShapeFrame::getCollisionAddon(const bool createIfNull)
+{
+  CollisionAddon* collisionData = getCollisionAddon();
+
+  if (createIfNull && nullptr == collisionData)
+    return createCollisionAddon();
+
+  return collisionData;
+}
+
+//==============================================================================
+DynamicsAddon* ShapeFrame::getDynamicsShapeNode(const bool createIfNull)
+{
+  DynamicsAddon* dynamicsData = getDynamicsAddon();
+
+  if (createIfNull && nullptr == dynamicsData)
+    return createDynamicsAddon();
+
+  return dynamicsData;
+}
```

Comment:
It would be nicer to replace these three functions with a generic `createOrGet<AddonT>` template function on `AddonManager`. Is there a strong reason to define them here?